File: rust/src/grid.rs
use std::fmt;

use crate::cell::Cell;
use image::{Rgb, RgbImage};
use imageproc::drawing::draw_line_segment_mut;

//trait Maze {
//
//}

#[derive(Clone, Debug)]
pub struct Grid {
    pub rows: i32,
    pub columns: i32,
    pub grid: Vec<Vec<Cell>>,
}

impl Grid {
    pub fn new(rows: i32, columns: i32) -> Self {
        let ret = Self {
            rows,
            columns,
            grid: prepare_grid(rows, columns),
        };
        configure_grid(&ret)
    }

    pub fn cell_by_id_mut(&mut self, id: i32) -> Option<&mut Cell> {
        for row in self.grid.iter_mut() {
            for cell in row.iter_mut() {
                if cell.id == id {
                    return Some(cell);
                }
            }
        }
        None
    }

    // This exists as a method on Cell in the Ruby and in my TypeScript
    // Rust was having none of that nonsense and I got mired in lifetime hell
    // Much easier to just elevate it to the grid level
    /// Links the two cells specified
    pub fn link(&mut self, origin: i32, target: i32) {
        // add target to origin
        self.cell_by_id_mut(origin)
            .unwrap_or_else(|| panic!("Tried to link nonexisting cell"))
            .links
            .push(target);
        // add origin to target
        self.cell_by_id_mut(target)
            .unwrap_or_else(|| panic!("Tried to link nonexisting cell"))
            .links
            .push(origin);
    }

    // saves an image of your maze, default cell size of 10
    pub fn to_img(&self, cell_size_param: Option<i32>) {
        let cell_size = cell_size_param.unwrap_or(10);
        let img_width = cell_size * self.columns;
        let img_height = cell_size * self.rows;
        let bg_color = Rgb([255, 255, 255]);
        let wall_color = Rgb([0, 0, 0]);

        let mut imgbuf = RgbImage::new((img_width + 1) as u32, (img_height + 1) as u32);

        // set all pixels to white
        for (_, _, px) in imgbuf.enumerate_pixels_mut() {
            *px = bg_color;
        }

        map_cells(self, |cell| {
            let x1 = (cell.column * cell_size) as f32;
            let y1 = (cell.row * cell_size) as f32;
            let x2 = ((cell.column + 1) * cell_size) as f32;
            let y2 = ((cell.row + 1) * cell_size) as f32;

            if cell.north.is_none() {
                draw_line_segment_mut(&mut imgbuf, (x1, y1), (x2, y1), wall_color)
            }
            if cell.west.is_none() {
                draw_line_segment_mut(&mut imgbuf, (x1, y1), (x1, y2), wall_color)
            }
            if !cell.linked(cell.east) {
                draw_line_segment_mut(&mut imgbuf, (x2, y1), (x2, y2), wall_color)
            }
            if !cell.linked(cell.south) {
                draw_line_segment_mut(&mut imgbuf, (x1, y2), (x2, y2), wall_color)
            }
        });

        imgbuf.save("maze.png").unwrap();
    }
}

impl<'a> fmt::Display for Grid {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut ret = String::new();

        // print top row
        ret.push_str("+");
        for _ in 0..self.columns {
            ret.push_str("---+")
        }
        ret.push_str("\n");

        // print each row - each grid row has 2 lines of text to print
        // "top" is our vertical walls "|"
        // "bottom" is our horizontal walls "---"
        for row in &self.grid {
            let mut top = String::new();
            top.push_str("|");
            let mut bottom = String::new();
            bottom.push_str("+");

            for cell in row {
                top.push_str("   ");
                if cell.linked(cell.east) {
                    top.push_str(" ");
                } else {
                    top.push_str("|");
                }

                if cell.linked(cell.south) {
                    bottom.push_str("   ");
                } else {
                    bottom.push_str("---");
                }
                bottom.push_str("+");
            }

            ret.push_str(&top);
            ret.push_str("\n");
            ret.push_str(&bottom);
            ret.push_str("\n");
        }

        write!(f, "{}", ret)
    }
}

// in Ruby, these were methods on grid
// I think this is going to eventually be a trait
// it sounds like it would need to be swappable

// initializes a prepared grid with neighbors
fn configure_grid(grid: &Grid) -> Grid {
    let mut ret = grid.clone();
    map_cells_mut(&mut ret, |cell: &mut Cell| {
        let row = cell.row;
        let col = cell.column;

        cell.north = get(&grid, row - 1, col);
        cell.south = get(&grid, row + 1, col);
        cell.west = get(&grid, row, col - 1);
        cell.east = get(&grid, row, col + 1);
    });
    ret
}

// instead of overloading []
/// Performs bounds-checking lookup on a &Grid
fn get(grid: &Grid, row: i32, column: i32) -> Option<i32> {
    if (row < 0 || row > grid.rows - 1) || (column < 0 || column > grid.columns - 1) {
        None
    } else {
        Some(grid.grid[row as usize][column as usize].id)
    }
}

/// takes a function to perform on each row, immutable cells
pub fn map_rows<F>(grid: &Grid, mut step: F)
where
    F: FnMut(&[Cell]),
{
    for row in &grid.grid {
        step(row);
    }
}

/// takes a function to perform on each cell, immutable cells
pub fn map_cells<F>(grid: &Grid, mut step: F)
where
    F: FnMut(&Cell),
{
    map_rows(grid, |row| {
        for cell in row {
            step(cell);
        }
    });
}

/// takes a function to perform on each row, mutable cells
fn map_rows_mut<F>(grid: &mut Grid, mut step: F)
where
    F: FnMut(&mut [Cell]),
{
    for row in &mut grid.grid {
        step(row);
    }
}

/// takes a function to perform on each cell, mutable cells
pub fn map_cells_mut<F>(grid: &mut Grid, mut step: F)
where
    F: FnMut(&mut Cell),
{
    map_rows_mut(grid, |row| {
        for cell in row.iter_mut() {
            step(cell);
        }
    });
}

/// returns a Vec<Vec<Cell>> of the given dimensions
fn prepare_grid(rows: i32, columns: i32) -> Vec<Vec<Cell>> {
    let mut ret = Vec::new();
    let mut current_id = 0;
    for i in 0..rows {
        let mut row = Vec::new();
        for j in 0..columns {
            row.push(Cell::new(current_id, i, j));
            current_id += 1;
        }
        ret.push(row);
    }
    ret
}

File: ts/src/grid.ts
import cell from './cell';

// To approximate img.line() from chunky_png in Ruby
const drawLine = (
  x1: number, y1: number, x2: number, y2: number, color: string, ctx: CanvasRenderingContext2D) => {
  ctx.beginPath();
  ctx.moveTo(x1, y1);
  ctx.lineTo(x2, y2);

  ctx.strokeStyle = color;
  ctx.stroke();
};

export default class Grid {
  private readonly context: CanvasRenderingContext2D;
  private readonly canvasW: number;
  private readonly canvasH: number;
  private grid: cell[][];
  constructor(
    private readonly canvas: HTMLCanvasElement,
    private readonly rows: number,
    private readonly columns: number,
    private readonly cellSize: number) {
    // set up canvas
    this.context = this.canvas.getContext('2d')!;
    this.canvasW = (rows * cellSize) + 1;
    this.canvasH = (columns * cellSize) + 1;
    this.canvas.width = this.canvasW;
    this.canvas.height = this.canvasH;

    // set up grid
    this.grid = [];
    this.prepareGrid();
    this.configureGrid();
  }

  private configureGrid = () => {
    this.eachCell((cell) => {
      const row = cell.row;
      const col = cell.column;

      cell.north = this.getCell(row - 1, col);
      cell.south = this.getCell(row + 1, col);
      cell.west = this.getCell(row, col - 1);
      cell.east = this.getCell(row, col + 1);
    });
  }

  public draw = (wallColor = '#000') => {
    this.context.clearRect(0, 0, this.canvasW, this.canvasH);

    this.eachCell((cell) => {
      // get boundaries
      const x1 = cell.column * this.cellSize;
      const y1 = cell.row * this.cellSize;
      const x2 = (cell.column + 1) * this.cellSize;
      const y2 = (cell.row + 1) * this.cellSize;

      // always draw north wall if we're in the top row
      if (!cell.north) { drawLine(x1, y1, x2, y1, wallColor, this.context); }

      // always draw west wall if we're in the west row
      if (!cell.west) { drawLine(x1, y1, x1, y2, wallColor, this.context); }

      // draw east wall unless linked
      if (!cell.isLinked(cell.east)) { drawLine(x2, y1, x2, y2, wallColor, this.context); }

      // draw south wall unless linked
      if (!cell.isLinked(cell.south)) { drawLine(x1, y2, x2, y2, wallColor, this.context); }
    });
  }

  public eachCell = (f: (cell: cell) => void) => {
    this.eachRow(row => row.forEach(cell => f(cell)));
  }

  public eachRow = (f: (row: cell[]) => void) => {
    this.grid.forEach(row => f(row));
  }

  private getCell = (row: number, col: number): cell | null => {
    if (row < 0 || row >= this.rows) { return null; }
    if (col < 0 || col >= this.columns) { return null; }
    return this.grid[row][col];
  }

  private prepareGrid = () => {
    for (let i = 0; i < this.rows; i += 1) {
      const row = [];
      for (let j = 0; j < this.columns; j += 1) {
        row.push(new cell(i, j));
      }
      this.grid.push(row);
    }
  }

  public toString = (): string => {
    let output = '+';
    for (let i = 0; i < this.columns; i += 1) {
      output += '---+';
    }
    output += '\n';
    this.grid.forEach((row) => {
      let top = '|';
      let bottom = '+';

      row.forEach((cell) => {
        const body = ` ${cell.contents()} `;
        const eastBoundary = (cell.isLinked(cell.east) ? ' ' : '|');
        top += body;
        top += eastBoundary;

        const southBoundary = (cell.isLinked(cell.south) ? '   ' : '---');
        const corner = '+';
        bottom += southBoundary;
        bottom += corner;
      });

      output += top;
      output += '\n';
      output += bottom;
      output += '\n';
    });
    return output;
  }
}